apps/emqx_dashboard/src/emqx_dashboard_swagger.erl

```diff
@@ -0,0 +1,228 @@
+        responses => Responses},
+        lists:usort(Refs1 ++ Refs2 ++ Refs3)
+    }.
+
+parameters(Params) ->
+    lists:foldl(fun({Name, Schema}, {Acc, RefsAcc}) ->
+        #{in := In, type := HoconType} = Schema,
+        {Type, Refs} = hocon_schema_to_spec(HoconType),
+        Spec0 = maps:without([type, desc], Schema),
+        Spec1 = Spec0#{schema => Type, name => Name},
+        Spec2 = add_required_to_path(In, Spec1),
+        Spec3 = trans_desc(Schema, Spec2),
+        {[Spec3 | Acc], Refs ++ RefsAcc}
+                end, {[], []}, Params).
+
+add_required_to_path(path, Spec) -> Spec#{required => true};
+add_required_to_path(_, Spec) -> Spec.
+
+trans_desc(#{desc := Desc}, Spec) -> Spec#{description => Desc};
+trans_desc(_, Spec) -> Spec.
+
+%% only object
+requestBody(Request) when is_map(Request) ->
+    case parse_object(Request) of
+        {[], Refs} -> {[], Refs};
+        {Props, Refs} ->
+            {#{<<"content">> =>
+            #{<<"application/json">> => #{<<"schema">> => Props}}}, Refs}
+    end.
+
+
+responses(Responses) ->
+    maps:fold(fun(Status, Hocon, {Acc, RefsAcc}) ->
+        if is_map(Hocon) ->
+            case maps:find(type, Hocon) of
+                {ok, HoconType} ->
+                    {Spec, Refs} = hocon_schema_to_spec(HoconType),
+                    Init0 = maps:without([type, desc], Hocon),
+                    Init = trans_desc(Hocon, Init0),
+                    Content = #{<<"application/json">> => #{<<"schema">> => Spec}},
+                    {Acc#{integer_to_binary(Status) => Init#{<<"content">> => Content}}, Refs ++ RefsAcc};
+                error -> %% object
+                    {Props, Refs} = parse_object(Hocon),
+                    Object = #{<<"content">> => #{<<"application/json">> => #{<<"schema">> => Props}}},
+                    {Acc#{integer_to_binary(Status) => Object}, Refs ++ RefsAcc}
+            end;
+            is_binary(Hocon) ->
+                {Acc#{integer_to_binary(Status) => #{description => Hocon}}, RefsAcc}
+        end
+              end, {#{}, []}, Responses).
+
+components(Refs) ->
+    maps:fold(fun(K, V, Acc) -> [#{K => V} | Acc] end, [], components(Refs, #{}, [])).
+
+components([], SpecAcc, []) -> SpecAcc;
+components([], SpecAcc, SubRefAcc) -> components(SubRefAcc, SpecAcc, []);
+components([{Module, Field} | Refs], SpecAcc, SubRefsAcc) ->
+    Props = apply(Module, fields, [Field]),
+    case is_list(Props) of
+        true -> %% object
+            {Object, SubRefs} = parse_object(maps:from_list(Props)),
+            NewSpecAcc = SpecAcc#{?REF(Module, Field) => Object},
+            components(Refs, NewSpecAcc, SubRefs ++ SubRefsAcc);
+        false -> %% schema
+            Init0 = maps:without([type, desc], Props),
+            Init = trans_desc(Props, Init0),
+            {Spec, SubRef} = hocon_schema_to_spec(maps:get(type, Props)),
+            NewSpecAcc = SpecAcc#{?REF(Module, Field) => maps:merge(Spec, Init)},
```

Comment:
I'd propose to respect the opaqueness of hocon schema, and be specific what we want to retrieve from the opaque structure.

meaning we should call `hocon_schema:field_schema/2` instead of `maps:get/2` to retrieve schema info.
e.g. to get type:
`hocon_schema:filed_schema(Props, type)`
to get default value:
`hocon_schema:field_schema(Props, default)`

Reply:
ok, Great to know hocon_schema/2